terraform/eval_diff_test.go

```diff
@@ -0,0 +1,78 @@
+package terraform
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestEvalFilterDiff(t *testing.T) {
+	ctx := new(MockEvalContext)
+
+	cases := []struct {
+		Node   *EvalFilterDiff
+		Input  *InstanceDiff
+		Output *InstanceDiff
+	}{
+		// With no settings, it returns an empty diff
+		{
+			&EvalFilterDiff{},
+			&InstanceDiff{Destroy: true},
+			&InstanceDiff{},
+		},
+
+		// Destroy
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{Destroy: false},
+			&InstanceDiff{Destroy: false},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{Destroy: true},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Destroy: true,
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{},
+				},
+			},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{
+						RequiresNew: true,
+					},
+				},
+			},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{},
+				},
+			},
+			&InstanceDiff{Destroy: false},
+		},
+	}
+
+	for i, tc := range cases {
+		var output *InstanceDiff
+		tc.Node.Diff = &tc.Input
+		tc.Node.Output = &output
+		if _, err := tc.Node.Eval(ctx); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+
+		if !reflect.DeepEqual(output, tc.Output) {
+			t.Fatalf("bad: %d\n\n%#v", i, output)
```

Comment:
> I figured that the tc.Input and output give you everything you need to fix the bug

Thing is, you don't get input and output... you get something like:

```
--- FAIL bad: 2

&InstanceDiff{},
```

As a future human encountering this, I know I broke something, but how am I to go about figuring out what I broke?

If instead, the output was

```
--- FAIL "preserves Destroy: false"

expected: &InstanceDiff{Destroy: false},
got: &InstanceDiff{},
```

I immediately know what I broke.

I am attempting to optimize for Cost for Future Human Reloading of this Context into their Brains (CFHRCB). :grinning: